src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,56 @@ public override bool Execute()
                 var filePath = Uri.Substring(FileUriProtocol.Length);
                 Log.LogMessage($"Copying '{filePath}' to '{DestinationPath}'");
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
-            else
-            {
-                Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
 
+            Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
+
+            Random rng = new Random();
+            for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+            {
                 using (var httpClient = new HttpClient())
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
-
                     try
                     {
+                        var httpResponse = await httpClient.GetAsync(Uri);
+
+                        if (httpResponse.StatusCode == HttpStatusCode.NotFound)
+                        {
+                            if (!string.IsNullOrEmpty(PrivateUri))
+                            {
+                                httpResponse = await httpClient.GetAsync(PrivateUri);
```

Comment:
This flow is a little weird with retry logic... it seems to mean *both* requests have to succeed in a single retry attempt to actually go through, but ideally it should only retry the private URI once it's established the public one is a 404.

I also wonder if each one should have its own retry count, so that failure to request the public URI doesn't reduce the number of chances it has to request the private one.

That leads me to think that maybe the request/copy logic should move into a different method, and this one should have two retry loops: one tries to download from the public blob storage, the second tries to download from private storage if the first found a 404.

This also makes it easy to print out an error that's specific to the endpoint being used, rather than the unclear `'{Uri}'/'{PrivateUri}'` below.

I don't know if this will ever matter in practice, just a suggestion. Sorry I didn't catch what was happening here in earlier iterations.

Reply:
Done.